text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,723 @@
+
+```rust
+fn foo(x: i32) -> i32 {
+    if x % 2 {
+        let x = x / 2;
+        // one branch uses `become`
+        become foo(x);
+    } else {
+        let x = x + 3;
+        // the other does not
+        foo(x) // == return foo(x);
+    }
+}
+```
+
+### Pitfall 3: Alternating `become` and `return` calls still grows the stack.
+([original example](https://github.com/rust-lang/rfcs/pull/1888#issuecomment-279062656))
+
+Here one function uses `become` the other `return`, this is another potential source of confusion. This mutual recursion
+would eventually overflow the stack. As mutual recursion can also happen across more functions, `become` needs to be
+used consistently in all functions if TCE is desired. (Maybe it is also possible to create a lint for these
+use cases as well.)
+
+```rust
+fn foo(n: i32) {
+    // oops, we forgot become ..
+    return bar(n); // or alternatively: `bar(n)`
+}
+
+fn bar(n: i32) {
+    become foo(n);
+}
+```
+
+### Pitfall 4: Caller location aka [`panic::Location::caller`](https://doc.rust-lang.org/std/panic/struct.Location.html#method.caller) and `#[track_caller]`
+
+Since `#[track_caller]` adds an argument, `#[track_caller]` functions can only tail call other `#[track_caller]` functions. So the following example will not compile.
+
+```rust
+use std::panic::Location;
+
+/// Returns the [`Location`] at which it is called.
+#[track_caller]
+fn get_caller_location() -> &'static Location<'static> {
+    Location::caller()
+}
+
+/// Returns a [`Location`] from within this function's definition.
+fn get_just_one_location() -> &'static Location<'static> {
+    become get_caller_location();
+}
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+<!-- This is the technical portion of the RFC. Explain the design in sufficient detail that:
+
+- Its interaction with other features is clear.
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+
+The section should return to the examples given in the previous section, and explain more fully how the detailed proposal makes those examples work. -->
+Implementation of this feature requires checks that all prerequisites to guarantee TCE are fulfilled.
+These checks are:
+
+- The `become` keyword is only used in place of `return`. The intent is to reuse the semantics of a `return` signifying "the end of a function". See the section on [tail-call-elimination](#tail-call-elimination) for examples.
+- The argument to `become` is a function (or method) call, that exactly matches the function signature and calling convention of the callee. The intent is to ensure a matching ABI. Note that mutability and lifetimes may differ as long as they pass borrow checking.
```

Comment:
Can you leave mutability as a future possibility/unresolved question? Allowing coercing mutability is somewhat annoying impl vise.

Reply:
updated here: https://github.com/rust-lang/rfcs/pull/3407/commits/12c495b3a86326d0b4e0ece4bf5e72a04975757a

Added as an unresolved question as I think this needs a bit more thought on how important it is. While adding `mut` to all arguments seems similar to requiring that all arguments match, it feels like that approach clashes a bit with allowing variable lifetimes.

I would be interested the general sentiment on this!

Reply:
I don't think mutability coercion should be allowed. My preferred way to express the "modulo lifetimes" issue is that the *function pointer* type of the called function must be coercible to the function pointer for the current function (or possibly vice versa, I didn't think too hard about the variance requirements here). That automatically ensures ABI compatibility while still allowing a lot of lifetime variation. Changing `&T` <-> `&mut T` is not allowed in this context, in either the arguments or the return.

Reply:
> Added as an unresolved question as I think this needs a bit more thought on how important it is. While adding `mut` to all arguments seems similar to requiring that all arguments match, it feels like that approach clashes a bit with allowing variable lifetimes.

I'm not sure I understand... do you mean that the following is allowed:
```rust
// snippet 1
fn f(x: u8) { become g(x) }
fn g(mut x: u8) {}
```
Or this is allowed:
```rust
// snippet 2
fn f(x: &mut u8) { become g(x) }
fn g(x: &u8) {}
```

Snippet 1 seems uncontroversial, `mut` there is a property of the binding in the function body, not the function signature. Snippet 2 on the other hand is a lot more controversial, because it needs a guarantee that `fn(&mut u8)` and `fn(&u8)` are ABI compatible, which in practice seems to be true, but it's quite a thing to guarantee...

Reply:
I meant snippet 2, I agree that snippet 1 is uncontroversial. As for tail calls we only care about ABI compatibility it seemed like this mismatch might be possible to support. Though, it wasn't clear to me that it is not already guaranteed. I'll change it to a future possibility if this should ever be guaranteed.

Should we mention snippet 1 in the RFC? It seems obvious enough to not require a separate description.

Reply:
I don't think we need to mention snippet 1 in the RFC.

Reply:
added here (forgot to add it to the commit message): https://github.com/rust-lang/rfcs/pull/3407/commits/3010c2e68c568f6c8741814c20034a5a38086c0e

Reply:
> ```rust
> // snippet 2
> fn f(x: &mut u8) { become g(x) }
> fn g(x: &u8) {}
> ```

Doesn't auto-de/ref kick-in here? If it's desugared to `become g(&*x);` after tyck, it works for free according to this RFC.

Reply:
@oxalica the issue is the "same arguments" restriction, since `f` and `g` do not have the same signature this would likely be disallowed. Autoref/deref in arguments isn't a problem, that would just be a normal coercion and there are no restrictions on operations in the function arguments.

Reply:
Requiring `fn(&u8)` and `fn(&mut u8)` to have the same ABI would also preclude references to small types (without interior mutability) being passed by its content directly (other semantic issues notwithstanding).

Reply:
That already isn't possible due to pointer addresses being observable. If you can prove that the pointer address won't be observed, you can also prove the absence of `become` and only then do the `&T` -> `T` transformation.

Reply:
> That already isn't possible due to pointer addresses being observable.

I’m aware of that – that was the other semantic issue I was mentioning. Still, allowing snippet 2 above would make this already turbulent change even harder to introduce. Whether we should treat addressless references as a lost cause or determine that it has enough benefits to leave on the table is of course a matter of debate, and one that’s probably out of scope here.

Reply:
We could leave snippet 2 unsupported (for now?) and it would have a straightforward workaround:

```
fn f(x: &mut u8) { 
    let y = &*x;
    become g(y) 
}

fn g(x: &u8) {}
```

Reply:
> That already isn't possible due to pointer addresses being observable. If you can prove that the pointer address won't be observed, you can also prove the absence of `become` and only then do the `&T` -> `T` transformation.

As someone unfamiliar with that optimization, is this optimization only planned for static function calls? Otherwise, I'm not sure how one would prove the absence of `become` as the current RFC supports tail calling dynamic calls (or I'm misunderstanding something).

Reply:
If you can't prove that the address isn't observable (like in the case of a dynamic call), then you can't turn `&T` arguments into `T`. If you can prove that the address isn't observable, you can also prove the absence of `become`. As such the `&T` to `T` arg optimization shouldn't matter for `become` as the set of cases where you can prove that it is allowed fully contains the set of cases where `become` isn't used.

Reply:
Let we have also Snippet 3 and 4:
```rust
// snippet 3
fn f(x: &mut u8) { become g(x) }
fn g(x: &u8) { if .. {become g(x)}}

// snippet 4
fn f(x: u8) { if .. {become g(x) }}
fn g(mut x: u8) { if .. {become f(x) }}
```
What to do here?

Reply:
@VitWW the content of a function can not change the ABI of a function, so both snippets are equivalent to snippet 2 regarding this discussion. The "workaround" would need to be that the programmer adapts the function signature.